engine/apps/alerts/tasks/notify_ical_schedule_shift.py

```diff
@@ -183,160 +26,137 @@ def notify_ical_schedule_shift(schedule_pk):
-                next_shifts_result = get_next_shifts_from_ical(
-                    calendar,
-                    schedule,
-                    next_overrides_priority,
-                    days_to_lookup=days_to_lookup,
-                )
-                if next_overrides_priority == 0 and next_shifts_result:
-                    next_overrides_priority = (
-                        max([next_shifts_result[uid]["priority"] for uid in next_shifts_result]) + 1
-                    )
-
-                next_shifts.update(next_shifts_result)
-
-        recalculate_shifts_with_respect_to_priority(next_shifts)
-
-        # drop events that already started
-        drop = []
-        for uid, next_shift in next_shifts.items():
-            if now > next_shift["start"]:
-                drop.append(uid)
-        for item in drop:
-            next_shifts.pop(item)
-
-        next_shifts_from_ical = sorted(next_shifts.values(), key=lambda shift: shift["start"])
+    shift_changed, diff_shifts = calculate_shift_diff(current_shifts, prev_shifts)
 
-        upcoming_shifts = []
-        # Add the earliest next_shift
-        if len(next_shifts_from_ical) > 0:
-            earliest_shift = next_shifts_from_ical[0]
-            upcoming_shifts.append(earliest_shift)
-            # Check if there are next shifts with the same start as the earliest
-            for shift in next_shifts_from_ical[1:]:
-                if shift["start"] == earliest_shift["start"]:
-                    upcoming_shifts.append(shift)
-
-        empty_oncall = len(current_shifts) == 0
-        if empty_oncall:
-            schedule.empty_oncall = True
-        else:
-            schedule.empty_oncall = False
+    # Do not notify if there is no difference between current and previous shifts
+    if not shift_changed:
+        task_logger.info(f"No shift diff found for schedule {schedule_pk}, organization {schedule.organization_id}")
+        # If prev shifts were converted to a new format, update related field in db
+        if prev_shifts_updated:
             schedule.current_shifts = json.dumps(current_shifts, default=str)
+            schedule.save(update_fields=["current_shifts"])
+        return
 
-        schedule.save(update_fields=["current_shifts", "empty_oncall"])
-
-        if len(new_shifts) > 0 or empty_oncall:
-            task_logger.info(f"new_shifts: {new_shifts}")
+    # Get only new/changed shifts to send a reminder message.
+    new_shifts = sorted(diff_shifts, key=lambda shift: shift["start"])
+
+    # get days_to_lookup for next shifts
+    if len(new_shifts) != 0:
+        days_to_lookup = (new_shifts[-1]["end"].date() - now.date()).days + 1
+        days_to_lookup = max([days_to_lookup, MIN_DAYS_TO_LOOKUP_FOR_THE_END_OF_EVENT])
+    else:
+        days_to_lookup = MIN_DAYS_TO_LOOKUP_FOR_THE_END_OF_EVENT
+
+    datetime_end = now + datetime.timedelta(days=days_to_lookup)
+
+    next_shifts_unfiltered = schedule.final_events(now, datetime_end, False, False)
```

Comment:
Not now, but maybe we could consider having only [one call](https://github.com/grafana/oncall/pull/2710/files?diff=unified&w=0#diff-ea1a6a0a4efe49256b2c05a6f516f0978db29859e0400bd4b758abb5919c1118R47) to `final_events`?

Reply:
It would be great, but in this case we need to decide how to calculate days_to_lookup